### ISSUE TITLE
Some questions regarding embeddings with pyannote

### ISSUE THREAD

**Issue by bml1g12** [pyannote/pyannote-audio#469]
Thanks for the great work with this repo. Just wanted to ask you a couple of questions regarding embeddings:

https://github.com/pyannote/pyannote-audio/tree/develop/tutorials/models/speaker_embedding 

"This model reaches 4.4% EER after 220 epochs (approximately 3 days of training). Note that this is obtained by simply using cosine distance between average embedding. This can be further reduced by using a different backend."

- I am looking to get a great embedding for speaker verification; what did you mean by using a different "backend" here? Do you mean a different backbone architecture (what would you recommend?) or did you mean cosine distance is not the best strategy.  In your excellent recent publication ("Neural building blocks for speaker diarization") , I see you mention that some people quite religiously use an PLDA based approach over cosine distance, so would be interested on your thoughts here.

- From your experience do you have any tips on getting a good quality speaker verification - I saw you mentioned VoxCeleb is advantageous over AMI for embedding due to more speakers.  And I think your paper is suggesting that TripletLoss outperforms AdditiveAngularMarginLoss used for the pretrained VoxCeleb model; so I guess I should try that. 

Thanks!

**Comment by hbredin** [comment]
> I am looking to get a great embedding for speaker verification; what did you mean by using a different "backend" here? Do you mean a different backbone architecture (what would you recommend?) or did you mean cosine distance is not the best strategy.  In your excellent recent publication ("Neural building blocks for speaker diarization") , I see you mention that some people quite religiously use an PLDA based approach over cosine distance, so would be interested on your thoughts here.

I am also looking for great embeddings :-) While I am happy with pyannote.audio segmentation (voice activity, speaker change, and overlapped speech detection), pyannote.audio embeddings still lag a bit behind state-of-the-art. @juanmc2005 did a great job comparing several approaches to training speaker embedding in his recent paper: [A Comparison of Metric Learning Loss Functions for End-to-End Speaker Verification](https://arxiv.org/abs/2003.14021). Slightly better pretrained pyannote.audio models are available in the [https://github.com/juanmc2005/SpeakerEmbeddingLossComparison](companion repository).

>  From your experience do you have any tips on getting a good quality speaker verification - I saw you mentioned VoxCeleb is advantageous over AMI for embedding due to more speakers.  And I think your paper is suggesting that TripletLoss outperforms AdditiveAngularMarginLoss used for the pretrained VoxCeleb model; so I guess I should try that.

Conclusions drawn by @juanmc2005 differ a bit on this point. I encourage you read the paper.

I'd be happy to collaborate on training great speaker embeddings. 
First thing would be to try and reproduce state-of-the-art approaches.
Let me know if you'd be interested in spending some time doing so...

**Comment by bml1g12** [comment]
Thanks for the link to the very informative paper; I am fairly new to the field of voice verification, so was useful to learn from the paper you linked about adaptive s-norm score normalisation also which might have a slight boosting effect (albeit the paper shows its not statistically significant for the additive angular margin loss function).

If I get time to work on it much, I will share any progress here. 

I'll try out the companion repo as a starting point, and also try as you suggested (https://github.com/pyannote/pyannote-audio/issues/471) using central cropping where possible over loose cropping. One thing I noticed is that we have a parameter `

**Comment by 1shershah** [comment]
https://github.com/pyannote/pyannote-audio/tree/develop/tutorials/models/speaker_embedding
Link is dead, can someone point me to the documentation that identifies the speaker ( seems default one is cosine similarity) ?
I believe the only option was to get the embedding from the segment and then use some clustering to see which embedding "cluster" close enough - so it was custom code on top of extracted embedding, does the library already have away to attach "speaker id" to the segments?
